internal/provider/resource_offer.go

```diff
@@ -2,173 +2,255 @@ package provider
+					stringplanmodifier.RequiresReplace(),
+				},
 			},
-			"url": {
+			"url": frameworkResSchema.StringAttribute{
 				Description: "The offer URL.",
-				Type:        schema.TypeString,
 				Computed:    true,
 			},
+			"id": frameworkResSchema.StringAttribute{
+				Computed: true,
+				PlanModifiers: []planmodifier.String{
+					stringplanmodifier.UseStateForUnknown(),
+				},
+			},
 		},
 	}
 }
 
-func resourceOfferCreate(ctx context.Context, d *schema.ResourceData, meta interface{}) diag.Diagnostics {
-	client := meta.(*juju.Client)
+func (o *offerResource) Create(ctx context.Context, req resource.CreateRequest, resp *resource.CreateResponse) {
+	// Prevent panic if the provider has not been configured.
+	if o.client == nil {
+		addClientNotConfiguredError(&resp.Diagnostics, "offer", "create")
+		return
+	}
 
-	var diags diag.Diagnostics
-	modelName := d.Get("model").(string)
-	modelInfo, err := client.Models.GetModelByName(modelName)
-	if err != nil {
-		return diag.FromErr(err)
+	var data offerResourceModel
+
+	// Read Terraform configuration from the request into the model
+	resp.Diagnostics.Append(req.Plan.Get(ctx, &data)...)
+	if resp.Diagnostics.HasError() {
+		return
 	}
 
+	modelName := data.ModelName.ValueString()
+	modelInfo, err := o.client.Models.GetModelByName(data.ModelName.ValueString())
+	if err != nil {
+		resp.Diagnostics.AddError("Client Error", fmt.Sprintf("Unable to resolve model UUID, got error: %s", err))
+		return
+	}
 	modelUUID := modelInfo.UUID
 	modelOwner := strings.TrimPrefix(modelInfo.OwnerTag, juju.PrefixUser)
 
-	//here we verify if the name property is set, if not set to the application name
-	var offerName string
-	name, ok := d.GetOk("name")
-	if ok {
-		offerName = name.(string)
-	} else {
-		offerName = d.Get("application_name").(string)
+	//here we verify if the name property is set, if not, set it to the application name
+	offerName := data.OfferName.ValueString()
+	if offerName == "" {
+		offerName = data.ApplicationName.ValueString()
 	}
 
-	result, errs := client.Offers.CreateOffer(&juju.CreateOfferInput{
+	response, errs := o.client.Offers.CreateOffer(&juju.CreateOfferInput{
 		ModelName:       modelName,
 		ModelUUID:       modelUUID,
 		ModelOwner:      modelOwner,
 		Name:            offerName,
-		ApplicationName: d.Get("application_name").(string),
-		Endpoint:        d.Get("endpoint").(string),
+		ApplicationName: data.ApplicationName.ValueString(),
+		Endpoint:        data.EndpointName.ValueString(),
 	})
 	if errs != nil {
-		if len(errs) == 1 {
-			return diag.FromErr(errs[0])
-		} else {
-			for _, v := range errs {
-				diags = append(diags, diag.FromErr(v)...)
-			}
-			return diags
+		for _, err := range errs {
+			resp.Diagnostics.AddError("Client Error", fmt.Sprintf("Unable to create offer, got error: %s", err))
 		}
+		return
+	}
+	tflog.Trace(ctx, fmt.Sprintf("create offer %q at %q", response.Name, response.OfferURL))
+
+	data.OfferName = types.StringValue(response.Name)
```

Comment:
given it's optional, should we be setting it by default? It is computed. If we add it here, the user will be required to add the optional value to their plan, afaik.

Reply:
not sure what you mean. Yes, it's a computed field, and this is where it's computed and set, just like the ID down below.